Connect DOAS to more ZoneHVAC Equipment - PTAC, PTHP, VRF, and UnitarySystem

Nigusse opened the pull request:
## Pull request overview

Added capability to connect zone HVAC equipment to central DOAS.  The following zone HVAC equipment were connected: 
       - ZoneHVAC:PackagedTerminalAirConditioner
       - ZoneHVAC:PackagedTerminalHeatPump
       - ZoneHVAC:TerminalUnit:VariableRefrigerantFlow
       - AirLoopHVAC:UnitarySystem ( as zone equipment)

Also combined two existing DOAS mixer objects `AirTerminal:SingleDuct:InletSideMixer` and `AirTerminal:SingleDuct:SupplySideMixer` into a single object `AirTerminal:SingleDuct:Mixer`.

This is a new feature and diffs are not expected.
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [x] CI status: all green or justified
- [x] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [x] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [x] Verify naming conventions and styles, memos and notes and defaults
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [x] If new idf included, locally check the err file and other outputs
- [x] Documentation changes in place

Nigusse committed:
NFP_DOAStoZoneHVACEquipment.md

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
updated NFP_DOAStoZoneHVACEquipment.md

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Design document and transition for get input.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Added design document

Nigusse committed:
code and IDD cleanup

Nigusse committed:
Added Reference Docs and code cleanup.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Documentation and IDD update

Nigusse committed:
Initial code change and Added unit test

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Added warning message and cleanup

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Reverted sizing changes and added Unitary System.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse commented:
This commit completes initial code change and unit tests.  The following zoneHVAC equipment are now connected to central DOAS:
       ZoneHVAC:PackagedTerminalAirConditioner
       ZoneHVAC:PackagedTerminalHeatPump
       ZoneHVAC:TerminalUnit:VariableRefrigerantFlow
       AirLoopHVAC:UnitarySystem (as a zone equipment)

The CI machine is expected to come all green.  This branch is ready for review.

Nigusse commented:
This branch is ready for review.  Any one interested in reviewing this new feature?

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39703,56 +39739,61 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
    A3, \field ZoneHVAC Terminal Unit Name
        \required-field
        \type object-list
        \object-list DOAToZonalUnit
```
@Nigusse If I'm following correctly, \reference DOAToZonalUnit needs to be added to each of the new equipment types (PTAC, PTHP, etc.)

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39703,56 +39739,61 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
+       \memo a dedicated outdoor air system (DOAS).
    A1, \field Name
        \required-field
    A2, \field ZoneHVAC Terminal Unit Object Type
```
@Nigusse "Terminal Unit" is confusing here in this field name and the next.  How about "ZoneHVAC Unit Object Type" and "ZoneHVAC Unit Object Name".

Nigusse replied:
Modified per the comment.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39703,56 +39739,61 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
        \type object-list
        \object-list DOAToZonalUnit
+       \note The name of ZoneHVAC equipment to which this terminal mixer will be connected.   
    A4, \field Terminal Unit Outlet Node Name
```
@Nigusse And I might consider changing the three "Terminal Unit . . . Node Name" fields to be "Mixer . . . Node Name"

Nigusse replied:
Modified per the comment.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39703,56 +39739,61 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
    A4, \field Terminal Unit Outlet Node Name
        \required-field
        \type node
+       \note The name of the primary air (outdoor air) inlet node of the mixer. This will be 
```
These notes belong with the Primary Air Inlet Node Name.

Nigusse replied:
Corrected per the comment. I found out this is different from the inputoutput reference document. Now they are consistent.

mjwitte commented:
@Nigusse At this point I've reviewed only the IDD.

Nigusse committed:
Updated Idd and documentation based on review comments

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse commented:
Updated IDD and Inputoutput documents based on the review feedback. @mjwitte Thanks!

Nigusse committed:
Added example files and cleaned up documentation.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Reference documentation cleanup.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse committed:
Added transition rule file.

Nigusse committed:
Updated AirTerminal:SingleDuct:Mixer field names in example files

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Conflicts:
	src/Transition/InputRulesFiles/Rules8-5-0-to-8-6-0.xls

Nigusse committed:
Updated transition Fortran Code.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse commented:
Added transition rules to the Fortran code. The CI machine is all green. This branch is ready for review and merging into develop.

nrel-bot-2 commented:
@Nigusse @lgentile it has been 15 days since this pull request was last updated.

mjwitte commented on testfiles/CMakeLists.txt:
```diff
@@ -328,6 +328,8 @@ ADD_SIMULATION_TEST(IDF_FILE PlateHeatExchanger.idf EPW_FILE USA_FL_Tampa.Intl.A
 ADD_SIMULATION_TEST(IDF_FILE Plenum.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE PlenumwithRetAirHeatGain.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE PondGroundHeatExchanger.idf EPW_FILE USA_AZ_Phoenix-Sky.Harbor.Intl.AP.722780_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE PTAC_ConnectedToDOAS.idf EPW_FILE USA_FL_Miami.Intl.AP.722020_TMY3.epw)
```
@Nigusse Minor quibble, we already have test files named DOAToFanCoilInlet.idf, DOAToFanCoilSupply.idf, etc.  Seems these new test files should follow the same naming pattern?

Nigusse replied:
@mjwitte Ok I can make those changes but note that each example file has a supply side and an inlet side connected zoneHVAC equipment. This was done to reduce the number of example file. So, these new example files could be renamed to: DOASToPTAC, DOASToPTHP, DOASToVRF, and DOASToUnitarySystem?

mjwitte replied:
Understood.  Note the old files are simpley "DOA" to . . ., not "DOAS".  Or we could rename the old ones, just seems the set should be consistent.

mjwitte commented:
@Nigusse There are some conflicts now with develop, primarily because the VRF terminal unit has some member functions now.  Shouldn't take long to resolve the conflicts to move your code to "this-->" form in a few places.

mjwitte commented on doc/input-output-reference/src/overview/group-zone-forced-air-units.tex:
```diff
@@ -3899,13 +3906,13 @@ \subsubsection{Inputs}\label{inputs-14-018}
+\paragraph{Field: Outdoor Air Mixer Object Type}\label{field-outdoor-air-mixer-object-type}
 
-This alpha field contains the identifying type of outside air mixer specified for the terminal unit. Outside air mixer type must be \textbf{OutsideAir:Mixer}. This field should be left blank when an outside air mixer is not simulated.
+This alpha field contains the identifying type of outdoor air mixer specified for the terminal unit. Outdoor air mixer type must be \textbf{OutdoorAir:Mixer}. This input field should be left blank when the VRF terminal unit is placed in \textbf{AirTerminal:SingleDuct:Mixer} object. If this field is left blank, an outdoor air mixer is not simulated.
```
@Nigusse For consistency with other similar text, "AirTerminal:SingleDuct:Mixer" should be italics here instead of bold.  Same in line 3915 below.

Nigusse replied:
edit as suggested in the comment.

Nigusse committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into DOASToZoneHVAC

Conflicts:
	src/EnergyPlus/HVACVariableRefrigerantFlow.cc
	testfiles/CMakeLists.txt

Nigusse committed:
Merge cleanup, idf rename and documents update.

Nigusse commented:
@mjwitte Merge cleanup index were replaced with "this ->", IO ref document updated as suggested and example idfs renamed.

mjwitte commented:
@Myoldmopar Why are there no windows CI reports here?

Myoldmopar commented:
I think I have the only Windows machine running right now.  It went down last week due to forced system updates, but I ran an hour worth of updates and reboots yesterday to get it going.  Then yesterday after I left work, apparently my UPS started beeping and so I think they may have just forced the UPS (and machine) off.  So I think it's down.  I'm working remote for the next few minutes but I'll be in the office to get it back up shortly.

mjwitte commented on testfiles/CMakeLists.txt:
```diff
@@ -330,6 +330,8 @@ ADD_SIMULATION_TEST(IDF_FILE PlateHeatExchanger.idf EPW_FILE USA_FL_Tampa.Intl.A
 ADD_SIMULATION_TEST(IDF_FILE Plenum.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE PlenumwithRetAirHeatGain.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 ADD_SIMULATION_TEST(IDF_FILE PondGroundHeatExchanger.idf EPW_FILE USA_AZ_Phoenix-Sky.Harbor.Intl.AP.722780_TMY3.epw)
+ADD_SIMULATION_TEST(IDF_FILE DOAToPTAC.idf EPW_FILE USA_FL_Miami.Intl.AP.722020_TMY3.epw)
```
@Nigusse Thanks for renaming these.  The new rows need to be moved no to keep this cmakelist in alphabetical order.  But before that, we need to decide if we want to keep all of these.  It's fantastic that we have a test file for each combination of DOA mixer and zoneHVAC type, but this now gives us 8 DOATo\* test files that all get packaged in ExampleFiles.  

@Myoldmopar Do we want to move some of these to _DOATo\* to reduce the number of ExampleFiles?

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39706,56 +39745,60 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
+       \note This is the outlet air node name of the mixer. This will be an inlet air node name 
+       \node of the ZoneHVAC equipment if the connection type in the input field Mixer Connection 
+       \note Type below is InletSide, else this will be inlet air node name of the conditioned zone
+       \node if the connection type in the input field Mixer Connection Type below is SupplySide.
```
\node --> \note, same 2 lines above

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -32868,15 +32869,17 @@ ZoneHVAC:PackagedTerminalAirConditioner,
         \key OutdoorAir:Mixer
         \note currently only one OutdoorAir:Mixer object type is available.
+        \note This field should be left blank, if the PTAC is connected to central
+        \note dedicated outdoor air through air terminal single duct mixer object.
```
1. Nit-picking - there shouldn't be a comma here " . . . left blank, if the . . .".  
2. And I think it would be clearer to say "through an AirTerminal:SingleDuct:Mixer object."
3. More nit-picking, in the existing note "currently only one . . . " please capitalize "Currently".
4. Please add these same comments to the fancoil and watertoairheatpump.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39706,56 +39745,60 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
        \required-field
        \type node
-   A5, \field Terminal Unit Primary Air Inlet Node Name
+       \note This is the outlet air node name of the mixer. This will be an inlet air node name 
```
This will be _the_ inlet air node name

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39706,56 +39745,60 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
-   A5, \field Terminal Unit Primary Air Inlet Node Name
+       \note This is the outlet air node name of the mixer. This will be an inlet air node name 
+       \node of the ZoneHVAC equipment if the connection type in the input field Mixer Connection 
+       \note Type below is InletSide, else this will be inlet air node name of the conditioned zone
```
else this will be _an_ inlet air node name . . .

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39706,56 +39745,60 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
        \type node
-   A6; \field Terminal Unit Secondary Air Inlet Node Name
+       \note The primary air (treated outdoor air) inlet node name of the mixer. This will be an 
+       \note outlet air node name of an AirLoopHVAC:ZoneSplitter, providing the connection to the 
```
of an AirLoopHVAC:ZoneSplitter _or AirLoopHVAC:SupplyPlenum_

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -39706,56 +39745,60 @@ AirTerminal:SingleDuct:ConstantVolume:CooledBeam,
-   A2, \field ZoneHVAC Terminal Unit Object Type
+       \note The secondary air (recirculating air) inlet node name of the mixer. This will be 
+       \note the outlet air node name of the ZoneHVAC equipment if the connection type in the 
+       \note input field mixer Connection Type below is SupplySide, or else this will be exhaust
```
this will be _an_ exhaust . . .
I realize these "the" and "an" changes seem picky, but these connections can be confusing, so making them as clear as possible is helpful.  In some cases it must be an exact match with a node name in the ZoneHVAC equipment, but in others, it's an arbitrary node name that also needs to be a zone inlet or zone exhaust.

mjwitte replied:
I haven't looked at the I/O Ref, but similar word changes may be needed there as well.

mjwitte commented:
- [x] The new test files should all be `Version,8.6`
- [x] _Page break issue, leave as-is._ The figures for AirTerminal:SingleDuct:Mixer are misplaced - they show up in the middle of the "Field: ZoneHVAC Unit Object Type" description
- [x] _Moved to #5603_  The figures for AirTerminal:SingleDuct:Mixer are missing lines.  This was there before, so it's not really a problem with this new feature, but if you could fix it here, that would be good. 
- [x] In the I/O ref section for AirTerminal:SingleDuct:Mixer there are several occurrences of "zoneHVAC", should be "ZoneHVAC".
- [x] Engineering Ref section for this also have some "zoneHVAC".
- [x] In the main description for AirTerminal:SingleDuct:Mixer, this long sentence can be broken into two "as a single stream to the conditioned zone at its inlet node or else if the AirTerminal:SingleDuct:Mixer object is connected". -->  ". . . its inlet node. If the AirTerminal . . ."
- [x] In the same paragraph, "specified by a user in the input field" --> "specified in the input field"
- [x] In the description for "Mixer Connection Type" in the sentence "Valid choices . . . " the choices are in italics, but should be bold to match the format later in the paragraph.
- [x] The fancoil I/O Ref fields for outdoor air mixer and outdoor air flow should get the same comments that you added to the new components about using with a DOAS mixer.
- [x] In the PTAC I/O Ref, "This input field should be left blank when the PTAC is placed in AirTerminal:SingleDuct:Mixer", "placed in" would be better as "connected to an".  Similar language occurs in other places.
- [x] While you're making changes, please fix AirloopHVAC:UnitarySystem heading for "Field: Supply Air Fan Operating Mode Schedule Name", it's at the wrong heading level.  It's interesting that AirloopHVAC:UnitarySystem does not have inputs for an outdoor air mixer, so it can't really be a fancoil or a ptac with it's own outdoor air.  That's something to consider adding - but I guess this feature works as-is.

mjwitte commented:
@Nigusse I've run the new test files as-is and as annual simulations.  They run without errors, but I haven't looked at results in detail yet.  I've added more doc comments [here](https://github.com/NREL/EnergyPlus/pull/5676#issuecomment-237900898).

Nigusse commented:
@mjwitte Thanks for the review. I will start addressing your review comments right away. Would be nice to know which example files to keep as internal.

mjwitte commented:
@Nigusse I would just reorder the list for now and keep all of the new files in the distribution.  That can be changed later if need be.

Nigusse commented:
@mjwitte I addressed all comments but these two:
- The figures for AirTerminal:SingleDuct:Mixer are misplaced - they show up in the middle of the "Field: ZoneHVAC Unit Object Type" description 
- The figures for AirTerminal:SingleDuct:Mixer are missing lines. 

Not sure what to change to avoid them. Any hint would help.

mjwitte commented:
I see, the figure position floats to the next page because there isn't room, and then LaTeX fills the prior page with later text, which normally would be a good thing, but here it lands in the middle of a list.  I don't know of a simple way to fix that, and it will move as soon as the document shifts with other changes.  So, never mind on that.
Regarding the lines on the figures, maybe @wfbuhl has the original artwork for those.  You could try to edit the png files, but should probably just let it rider for now.  I'll add to #5603

Nigusse committed:
Documents, Idd and CMakeLists cleanup.

Nigusse committed:
Merge branch 'develop' into DOASToZoneHVAC

Nigusse commented:
I see figures position floating in IORef for "GroundHeatExchanger:Surface" object as well.

mjwitte committed:
DOAS to ZoneHVAC minor doc and idd edits

mjwitte committed:
DOAS to ZoneHVAC more doc edits

mjwitte commented:
@Nigusse Tested transition. It modifies the AirTerminal:SingleDuct:*Mixer objects correctly, but it also needs to change the Air Terminal Object Type field in the corresponding ZoneHVAC:AirDistributionUnit object.  Be sure to pull the branch before making any changes to get the doc edits I added.

Nigusse committed:
Updated transition rule.

Nigusse committed:
Updated transition rule.

Nigusse committed:
Resolved conflict

Nigusse commented:
fixed incorrect syntax in the transition rules Fortran code that lead build to fail.

Nigusse committed:
Updated Rules8-5-0-to-8-6-0.xls file.

Myoldmopar committed:
Merge develop; resolve conflicts

Myoldmopar commented:
@Nigusse I merged develop to resolve the conflicts.  They were only in the transition source and related files.  I'll let CI go over this last commit, but I don't have anything to add here.  Thanks to @mjwitte for his review on it.  I'll merge as long as CI is happy.

I think it might be worth a heads up of this new capability for one @kbenne.
